crates/containerd-shim-wasm/src/libcontainer_instance/executor.rs

```diff
@@ -0,0 +1,78 @@
+use std::fs::File;
+use std::io::Read;
+use std::os::unix::prelude::PermissionsExt;
+
+use anyhow::{bail, ensure, Context};
+use libcontainer::workload::default::DefaultExecutor;
+use libcontainer::workload::{
+    Executor as LibcontainerExecutor, ExecutorError as LibcontainerExecutorError,
+    ExecutorValidationError,
+};
+use oci_spec::runtime::Spec;
+
+use super::context::RuntimeContext;
+use super::engine::Engine;
+use crate::sandbox::Stdio;
+
+#[derive(Clone)]
+pub(crate) struct Executor<E: Engine> {
+    pub engine: E,
+    pub stdio: Stdio,
+}
+
+impl<E: Engine> LibcontainerExecutor for Executor<E> {
+    fn validate(&self, spec: &Spec) -> Result<(), ExecutorValidationError> {
```

Comment:
ummm it feels strange to me that `validate` is already checking either Linux binary or wasm bianry but `execute` has to check the same thing again. This time, it delegates to different runtime to exec. 

I guess this makes sense and preserves what we had before. It's just a bit efficient.

Reply:
See https://github.com/containerd/runwasi/pull/290#discussion_r1309404870
We could use internal mutability, and store the result of the check so that we don't have to repeat it.

Reply:
Could we mutate a state in `Executor` to save the `validate` result and reuse in `execute`? 

Something like `is_linux_executable: RefCell<bool>`

Reply:
Ah you commented a few seconds before me!

Reply:
Done :-)